### ISSUE TITLE
Implement Recover Block

### ISSUE THREAD

**Issue by teggotic** [jemc/mare#22]
pony has syntax for recovering rcaps
```pony
let s = recover iso String end
```

**Comment by jemc** [comment]
Yeah, so as a general statement I am looking at three main areas of work here:

---

:one:
Firstly, try to avoid the need for an explicit `recover` in as many cases as possible - for example, in the `recover iso String end` Pony code you gave, the Mare compiler should know that any call which has all sendable arguments (or no arguments, as in this case) can be trivially recovered with no need for a `recover` block, based on a constraint from the left side. In other words, I want this code to work with no need for a `recover` block:

```mare
string1 = String.new     # this is a String'ref - the normal result type for this call
string2 iso = String.new # this is a String'iso - lifted automatically from ref to iso
```

There are probably other cases where explicit recovery shouldn't need to be specified.

---

:two:
Tracking of "same origin" or "different origin" for an object in the compiler - the compiler should know that it is safe to use and combine non-sendable objects together as long as the compiler knows them to be from the "same origin", and that this can be represented as an `iso` (the origin) interacting with `box`es and `ref`s that "belong" to it, without necessarily needing a `recover` block.

Note that this is closer to how the Verona runtime does reference capabilities, and it is more permissive than the Pony model, while still maintaining the same concurrency guarantees, and being compatible with the Pony runtime.

More thought and design is needed - new syntax too. Lots of design work to do here.

---
:three:
If work from :one: and :two: is not sufficient to eliminate Pony-style explicit recover blocks, we should add them, taking care that they are analyzed by the compiler in an "origin aware" way as described in :two:, rather than following the same model the Pony compiler uses for verification.

**Comment by teggotic** [comment]
>There are probably other cases where explicit recovery shouldn't need to be specified.
I think when you pass something as a parameter
```mare
:fun foo(bar String'iso)
  None
:fun baz
  @foo(String.new) // or even "" because val should be recoverable to iso
```

**Comment by jemc** [comment]
Yeah, that kind of case, and others where the value is used somewhere an `iso` or `val` is required would all be covered by the same kind of compiler change.

**Comment by teggotic** [comment]
2nd is a bit hard for me to understand

**Comment by jemc** [comment]
Helpful reading for case :two: are these two comments from Paul in ponyc, in which he was brainstorming ideas that he later helped to incorporate into Verona:
- https://github.com/ponylang/ponyc/pull/2043#issuecomment-319835433
- https://github.com/ponylang/ponyc/pull/2043#issuecomment-319840093


My ideas of syntax are a bit different from what he was suggesting - he was suggested numbered regions, but I think it might be more convenient to talk about appending a prime to different capabilities.

For example, `ref` and `ref'` would be mutable objects from different origins, which would not be allowed to "touch" eachother, though you could take "sendable" things across the boundary between the two, and you could do this without necessarily needing a `recover` block for it.

Hopefully this syntax could be done in such a way that you only need to deal with "distinct origin" capability notation like `ref'` when you really need to deal with multiple origins.

**Comment by teggotic** [comment]
How do you distinguish origins

**Comment by jemc** [comment]
We should definitely not do this until we have better control flow analysis: https://github.com/jemc/mare/issues/75

Marking this ticket as blocked by that one.

**Comment by jemc** [comment]
Finished the blocking ticket: #75, so am removing the `BLOCKED` label.

However, this still needs lots of thought and design before beginning work.
